File: internal/domain/errors_test.go
package domain

import (
	"net/http"
	"testing"
)

func TestAPIError_Error(t *testing.T) {
	tests := []struct {
		name     string
		err      *APIError
		expected string
	}{
		{
			name:     "error with type and message",
			err:      &APIError{Type: ErrorTypeInvalidRequest, Message: "bad request"},
			expected: "invalid_request: bad request",
		},
		{
			name:     "error with type, code, and message",
			err:      &APIError{Type: ErrorTypeRateLimit, Code: ErrorCodeRateLimitExceeded, Message: "rate limited"},
			expected: "rate_limit (rate_limit_exceeded): rate limited",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := tt.err.Error(); got != tt.expected {
				t.Errorf("Error() = %q, want %q", got, tt.expected)
			}
		})
	}
}

func TestAPIError_HTTPStatusCode(t *testing.T) {
	tests := []struct {
		name     string
		err      *APIError
		expected int
	}{
		{
			name:     "invalid request",
			err:      &APIError{Type: ErrorTypeInvalidRequest},
			expected: http.StatusBadRequest,
		},
		{
			name:     "authentication error",
			err:      &APIError{Type: ErrorTypeAuthentication},
			expected: http.StatusUnauthorized,
		},
		{
			name:     "permission error",
			err:      &APIError{Type: ErrorTypePermission},
			expected: http.StatusForbidden,
		},
		{
			name:     "not found error",
			err:      &APIError{Type: ErrorTypeNotFound},
			expected: http.StatusNotFound,
		},
		{
			name:     "rate limit error",
			err:      &APIError{Type: ErrorTypeRateLimit},
			expected: http.StatusTooManyRequests,
		},
		{
			name:     "overloaded error",
			err:      &APIError{Type: ErrorTypeOverloaded},
			expected: http.StatusServiceUnavailable,
		},
		{
			name:     "server error",
			err:      &APIError{Type: ErrorTypeServer},
			expected: http.StatusInternalServerError,
		},
		{
			name:     "context length error",
			err:      &APIError{Type: ErrorTypeContextLength},
			expected: http.StatusBadRequest,
		},
		{
			name:     "max tokens error",
			err:      &APIError{Type: ErrorTypeMaxTokens},
			expected: http.StatusBadRequest,
		},
		{
			name:     "unknown error type",
			err:      &APIError{Type: ErrorType("unknown")},
			expected: http.StatusInternalServerError,
		},
		{
			name:     "explicit status code",
			err:      &APIError{Type: ErrorTypeInvalidRequest, StatusCode: http.StatusConflict},
			expected: http.StatusConflict,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := tt.err.HTTPStatusCode(); got != tt.expected {
				t.Errorf("HTTPStatusCode() = %d, want %d", got, tt.expected)
			}
		})
	}
}

func TestNewAPIError(t *testing.T) {
	err := NewAPIError(ErrorTypeInvalidRequest, "test message")
	if err.Type != ErrorTypeInvalidRequest {
		t.Errorf("Type = %v, want %v", err.Type, ErrorTypeInvalidRequest)
	}
	if err.Message != "test message" {
		t.Errorf("Message = %q, want %q", err.Message, "test message")
	}
}

func TestAPIError_WithCode(t *testing.T) {
	err := NewAPIError(ErrorTypeRateLimit, "rate limited").WithCode(ErrorCodeRateLimitExceeded)
	if err.Code != ErrorCodeRateLimitExceeded {
		t.Errorf("Code = %v, want %v", err.Code, ErrorCodeRateLimitExceeded)
	}
}

func TestAPIError_WithParam(t *testing.T) {
	err := NewAPIError(ErrorTypeInvalidRequest, "invalid value").WithParam("temperature")
	if err.Param != "temperature" {
		t.Errorf("Param = %q, want %q", err.Param, "temperature")
	}
}

func TestAPIError_WithStatusCode(t *testing.T) {
	err := NewAPIError(ErrorTypeInvalidRequest, "conflict").WithStatusCode(http.StatusConflict)
	if err.HTTPStatusCode() != http.StatusConflict {
		t.Errorf("HTTPStatusCode() = %d, want %d", err.HTTPStatusCode(), http.StatusConflict)
	}
}

func TestAPIError_WithSourceAPI(t *testing.T) {
	err := NewAPIError(ErrorTypeServer, "error").WithSourceAPI(APITypeOpenAI)
	if err.SourceAPI != APITypeOpenAI {
		t.Errorf("SourceAPI = %v, want %v", err.SourceAPI, APITypeOpenAI)
	}
}

func TestConvenienceConstructors(t *testing.T) {
	tests := []struct {
		name         string
		constructor  func(string) *APIError
		message      string
		expectedType ErrorType
		expectedCode ErrorCode
	}{
		{
			name:         "ErrInvalidRequest",
			constructor:  ErrInvalidRequest,
			message:      "bad request",
			expectedType: ErrorTypeInvalidRequest,
			expectedCode: "",
		},
		{
			name:         "ErrAuthentication",
			constructor:  ErrAuthentication,
			message:      "invalid key",
			expectedType: ErrorTypeAuthentication,
			expectedCode: "",
		},
		{
			name:         "ErrPermission",
			constructor:  ErrPermission,
			message:      "access denied",
			expectedType: ErrorTypePermission,
			expectedCode: "",
		},
		{
			name:         "ErrNotFound",
			constructor:  ErrNotFound,
			message:      "model not found",
			expectedType: ErrorTypeNotFound,
			expectedCode: "",
		},
		{
			name:         "ErrRateLimit",
			constructor:  ErrRateLimit,
			message:      "rate limited",
			expectedType: ErrorTypeRateLimit,
			expectedCode: ErrorCodeRateLimitExceeded,
		},
		{
			name:         "ErrOverloaded",
			constructor:  ErrOverloaded,
			message:      "server overloaded",
			expectedType: ErrorTypeOverloaded,
			expectedCode: "",
		},
		{
			name:         "ErrServer",
			constructor:  ErrServer,
			message:      "internal error",
			expectedType: ErrorTypeServer,
			expectedCode: "",
		},
		{
			name:         "ErrContextLength",
			constructor:  ErrContextLength,
			message:      "context too long",
			expectedType: ErrorTypeContextLength,
			expectedCode: ErrorCodeContextLengthExceeded,
		},
		{
			name:         "ErrMaxTokens",
			constructor:  ErrMaxTokens,
			message:      "max tokens exceeded",
			expectedType: ErrorTypeMaxTokens,
			expectedCode: ErrorCodeMaxTokensExceeded,
		},
		{
			name:         "ErrOutputTruncated",
			constructor:  ErrOutputTruncated,
			message:      "output was truncated",
			expectedType: ErrorTypeMaxTokens,
			expectedCode: ErrorCodeOutputTruncated,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			err := tt.constructor(tt.message)
			if err.Type != tt.expectedType {
				t.Errorf("Type = %v, want %v", err.Type, tt.expectedType)
			}
			if err.Code != tt.expectedCode {
				t.Errorf("Code = %v, want %v", err.Code, tt.expectedCode)
			}
			if err.Message != tt.message {
				t.Errorf("Message = %q, want %q", err.Message, tt.message)
			}
		})
	}
}

func TestAPIError_Chaining(t *testing.T) {
	err := NewAPIError(ErrorTypeInvalidRequest, "test").
		WithCode(ErrorCodeContextLengthExceeded).
		WithParam("messages").
		WithStatusCode(http.StatusBadRequest).
		WithSourceAPI(APITypeAnthropic)

	if err.Type != ErrorTypeInvalidRequest {
		t.Errorf("Type = %v, want %v", err.Type, ErrorTypeInvalidRequest)
	}
	if err.Code != ErrorCodeContextLengthExceeded {
		t.Errorf("Code = %v, want %v", err.Code, ErrorCodeContextLengthExceeded)
	}
	if err.Param != "messages" {
		t.Errorf("Param = %q, want %q", err.Param, "messages")
	}
	if err.StatusCode != http.StatusBadRequest {
		t.Errorf("StatusCode = %d, want %d", err.StatusCode, http.StatusBadRequest)
	}
	if err.SourceAPI != APITypeAnthropic {
		t.Errorf("SourceAPI = %v, want %v", err.SourceAPI, APITypeAnthropic)
	}
}

File: ts/packages/gateway-core/src/domain/errors.test.ts
import { describe, it, expect } from 'vitest';
import {
    APIError,
    errInvalidRequest,
    errAuthentication,
    errRateLimit,
    errServer,
    errNotFound,
    isAPIError,
    toOpenAIError,
    toAnthropicError,
} from './errors';

describe('APIError', () => {
    describe('constructor', () => {
        it('should create error with message', () => {
            const error = new APIError('invalid_request', 'Test error');
            expect(error.message).toBe('Test error');
            expect(error.type).toBe('invalid_request');
            expect(error.name).toBe('APIError');
        });

        it('should create error with options', () => {
            const error = new APIError('server', 'Server error', {
                code: 'model_not_found',
                param: 'model',
                statusCode: 503,
                sourceAPI: 'openai',
            });

            expect(error.type).toBe('server');
            expect(error.code).toBe('model_not_found');
            expect(error.param).toBe('model');
            expect(error.statusCode).toBe(503);
            expect(error.sourceAPI).toBe('openai');
        });

        it('should use default status code based on type', () => {
            expect(new APIError('invalid_request', 'test').statusCode).toBe(400);
            expect(new APIError('authentication', 'test').statusCode).toBe(401);
            expect(new APIError('permission', 'test').statusCode).toBe(403);
            expect(new APIError('not_found', 'test').statusCode).toBe(404);
            expect(new APIError('rate_limit', 'test').statusCode).toBe(429);
            expect(new APIError('server', 'test').statusCode).toBe(500);
            expect(new APIError('overloaded', 'test').statusCode).toBe(503);
        });
    });

    describe('builder methods', () => {
        it('should chain builder methods', () => {
            const error = new APIError('invalid_request', 'Test')
                .withCode('model_not_found')
                .withParam('model')
                .withStatusCode(404)
                .withSourceAPI('openai');

            expect(error.code).toBe('model_not_found');
            expect(error.param).toBe('model');
            expect(error.statusCode).toBe(404);
            expect(error.sourceAPI).toBe('openai');
        });
    });

    describe('toJSON', () => {
        it('should serialize to JSON format', () => {
            const error = new APIError('invalid_request', 'Invalid model', {
                code: 'model_not_found',
                param: 'model',
            });

            const json = error.toJSON();
            expect(json).toEqual({
                type: 'invalid_request',
                message: 'Invalid model',
                code: 'model_not_found',
                param: 'model',
            });
        });
    });
});

describe('convenience constructors', () => {
    it('errInvalidRequest should create invalid_request error', () => {
        const error = errInvalidRequest('Bad input');
        expect(error.type).toBe('invalid_request');
        expect(error.message).toBe('Bad input');
        expect(error.statusCode).toBe(400);
    });

    it('errAuthentication should create authentication error', () => {
        const error = errAuthentication('Invalid key');
        expect(error.type).toBe('authentication');
        expect(error.statusCode).toBe(401);
    });

    it('errRateLimit should create rate_limit error', () => {
        const error = errRateLimit('Too many requests');
        expect(error.type).toBe('rate_limit');
        expect(error.statusCode).toBe(429);
    });

    it('errServer should create server error', () => {
        const error = errServer('Internal error');
        expect(error.type).toBe('server');
        expect(error.statusCode).toBe(500);
    });

    it('errNotFound should create not_found error', () => {
        const error = errNotFound('Model not found');
        expect(error.type).toBe('not_found');
        expect(error.statusCode).toBe(404);
    });
});

describe('isAPIError', () => {
    it('should return true for APIError instances', () => {
        const error = new APIError('server', 'Test');
        expect(isAPIError(error)).toBe(true);
    });

    it('should return false for regular Error instances', () => {
        const error = new Error('Test');
        expect(isAPIError(error)).toBe(false);
    });

    it('should return false for non-error values', () => {
        expect(isAPIError(null)).toBe(false);
        expect(isAPIError(undefined)).toBe(false);
        expect(isAPIError('error')).toBe(false);
        expect(isAPIError({ message: 'test' })).toBe(false);
    });
});

describe('toOpenAIError', () => {
    it('should convert to OpenAI error format', () => {
        const error = new APIError('invalid_request', 'Invalid model', {
            code: 'model_not_found',
            param: 'model',
        });

        const openaiError = toOpenAIError(error);
        expect(openaiError).toEqual({
            error: {
                type: 'invalid_request_error',
                code: 'model_not_found',
                message: 'Invalid model',
                param: 'model',
            },
        });
    });
});

describe('toAnthropicError', () => {
    it('should convert to Anthropic error format', () => {
        const error = new APIError('authentication', 'Invalid API key');

        const anthropicError = toAnthropicError(error);
        expect(anthropicError).toEqual({
            type: 'error',
            error: {
                type: 'authentication_error',
                message: 'Invalid API key',
            },
        });
    });
});
